### ISSUE TITLE
Implementation types incorrectly included in overloaded function, causing unknowns

### ISSUE THREAD

**Issue by mikeshardmind** [microsoft/pyright#5007]
**Describe the bug**

A function typed with 3 overloads is generating an error showing 4 overloads, 1 of which appears to be pyright's inferred types for the untyped actual implementation.

Per [PEP 484](https://peps.python.org/pep-0484/#function-method-overloading)

> The @overload-decorated definitions are for the benefit of the type checker only, since they will be overwritten by the non-@overload-decorated definition, **while the latter is used at runtime but should be ignored by a type checker.**

emphasis mine

**To Reproduce**

Use a function with typed overloads and an untyped implementation in strict mode.

**Expected behavior**
The untyped actual implementation to be ignored.

**Screenshots or Code**

![image](https://user-images.githubusercontent.com/22035876/234273270-9e8c5a94-01e5-4bd7-8fe7-c061dfd8a841.png)

(there are only 3 overloads, nothing relevant is out of view)

Generates this error, containing 4 Overloads, the last of which matches pyright's inference of the untyped implementation.
```
Type of "decode" is partially unknown
  Type of "decode" is "Overload[(buf: bytes | str, *, dec_hook: ((type, Any) -> Any) | None = None) -> Any, (buf: bytes | str, *, type: Type[T@decode] = ..., dec_hook: ((type, Any) -> Any) | None = None) -> T@decode, (buf: bytes | str, *, type: Any = ..., dec_hook: ((type, Any) -> Any) | None = None) -> Any, (buf: Unknown, *, type: Any = Any, dec_hook: Unknown | None = None) -> (dict[str, Any] | Unknown | Any)]"
```

Using msgspec installed from source (unreleased other type fixes) specified as
```
msgspec @ git+https://github.com/jcrist/msgspec.git@dc885f96b14f116c1e116932d7120d5aa1ab7d3b
```

**VS Code extension or command-line**
via pylance v2023.4.30 in VS Code

**Additional context**

**Comment by erictraut** [comment]
Oh, I see what you mean. You appear to be using strict mode type checking. In strict mode, you must supply type annotations for the implementation as well.

Or is the overloaded function defined in code that is outside of your project?

**Comment by mikeshardmind** [comment]
The overloaded function is imported from msgspec and is not part of my code.

The following should cause it to be replicatable
```py
# pip install "msgspec @ git+https://github.com/jcrist/msgspec.git@dc885f96b14f116c1e116932d7120d5aa1ab7d3b"
import msgspec

class Config(msgspec.Struct):
    x: int


def parse_config(raw: bytes) -> Config:
    return msgspec.toml.decode(raw, type=Config)
```

msgspec implementation and typehints at that revision here: https://github.com/jcrist/msgspec/blob/dc885f96b14f116c1e116932d7120d5aa1ab7d3b/msgspec/toml.py#L87-L116

**Comment by mikeshardmind** [comment]
> In strict mode, you must supply type annotations for the implementation as well.
Is there any documentation for this? The relevant PEP notes that type checkers should ignore the implementation type and instead use the overloads.

**Comment by mikeshardmind** [comment]
This currently occurs in code using msgspec, but not a part of msgspec, so if that's the stance of pyright, then this is currently erroring in a way that is unintended.

Beyond this, the PEP for this quoted above clearly says that type checkers should ignore the implementation's typing. That said, I can understand a rationale that contradicts this being desirable as it should be possible to check (from a maintainer's POV) that an implementation that doesn't allow certain combinations (combinations not exposed via overloads) are still handled in the implementation and raised to the consumer. Is this divergence from the PEP worth documenting in a note about strict in configuration or would there be a better place for this (and should this be split into a separate issue from that this is showing up outside of that library?)

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.305, which I just published. It will also be included in a future release of pylance.
